ffmpeg export

yohannd1 opened the pull request:
**NOTE**: this PR is on hold. Many new audio formats are already supported on master so it's alright :)

---

An attempt at supporting custom format export via ffmpeg, using the fact that this command works:
```
cat abc.wav | ffmpeg -f wav -i pipe:0 -f mp3 out.mp3
```
The idea then is to open a `ffmpeg` child process and write the export data (which would be written to a `.wav` file) to it, without needing to link against ffmpeg's libraries. And it effectively makes it optional!

Progress:
- [x] Initial UI changes (https://github.com/tildearrow/furnace/pull/2309/commits/c03dd77975216d56b07e80e6478bd0d0c3d6af70)
- [x] Functional linux fork-pipe-dup2-exec implementation (https://github.com/tildearrow/furnace/pull/2309/commits/ef902091415c8001c1fb5e0976fa57c06bce3967)
- [x] Support custom flags
- [x] Basic format support (MP3/FLAC/OGG/M4A/OPUS)
- [x] Bug fix: export is at 2x!
- [x] Detect whether `ffmpeg` is in PATH
- [x] Setting for custom ffmpeg path
- [x] Writer API: remove `exportOutputs` arg (it should be kept in the writer struct itself, not received as an argument)
- [x] Make export work with the "multiple chips" mode
- [x] Make export work with the "multiple channels" mode
- [ ] Visual feedback on export errors
- [ ] File chooser for custom ffmpeg path setting
- [ ] Support Windows/Wine (MinGW)
- [ ] Support Windows (MSVC)
- [ ] Support MacOS
- [ ] Custom export commands (with different templated flags, e.g. `/path/to/ffmpeg -f pcm_s16le -i pipe:0 -f flac %{out_path} %{extra_flags}`)
- [ ] TODO check (`git diff tilde/master | rg '\b(TODO|FIXME|XXX)\b'`)

yohannd1 committed:
ffmpeg export: add format option in audio export (non functional)

yohannd1 committed:
ffmpeg export: four basic file formats (only visual for now)

yohannd1 committed:
ffmpeg export: Subprocess class (POSIX impl) (and general changes)

a functional usage example of the Subprocess class is currently at
src/engine/wavOps.cpp:124 (sorry, it's misplaced)

yohannd1 committed:
ffmpeg export: successfully pipe the exported WAV data

yohannd1 commented:
I'm kind of second-guessing this because I noticed `libsndfile` (already used) supports a few file formats, but I suppose `ffmpeg` has way more options so this is still useful.

yohannd1 committed:
ffmpeg export: implement Subprocess::wait

Oh my fucking god this was so goddamn hard to implement. But it's here.

I don't know how I was so lucky but I came across this:

https://stackoverflow.com/questions/53734902/why-child-process-is-waiting-while-using-fork-and-pipes-for-stdin-from-parents

which made me realize stuff can go bad if I do not close the file
descriptors at every possible moment. In this case, I had to close as
many unrelated ends of stdinPipe as possible so the subprocess could
figure out the export was finished

yohannd1 committed:
ffmpeg export: functional flac/ogg/mp3 options

yohannd1 committed:
ffmpeg export: minor refactoring, and some notes on what to do next

who knew IPC could be so hard

yohannd1 committed:
use lowerCaseCopy :)

yohannd1 committed:
ffmpeg export: actually notice when ffmpeg fails

This was a nightmare :)

I did learn how poll() works though. That's nice.

But I had to go through so much to find that out lol... an entire
afternoon.

yohannd1 committed:
ffmpeg export: save ffmpeg flags

yohannd1 committed:
ffmpeg export: (failed) attempt at feeding raw PCM into fmpeg (beware - loud noises!)

yohannd1 committed:
ffmpeg export: don't error on windows

the feature is still not implemented, but at least now it compiles with
a message saying it's not yet supported

yohannd1 committed:
ffmpeg export: still trying to make work in MSVC

it does on mingw though

yohannd1 committed:
cross-mingw build: detect whether there is a -posix suffix

LTVA1 commented:
The problem with `libsndfile` is that it's not easy to statically link `libflac`, `libogg`, `libvorbis` and whatever MP3 requires. I and others tried multiple times and it just doesn't compile. So maybe ffmpeg may be a workaround. You coul even try to enable Furnace to support other formats like flac and mp3 by issuing a command when user opens file so that ffmpeg converts it to wav which Furnace can work with

yohannd1 commented:
Oh, I understand, thanks!
As for the second part, do you mean while importing? I was thinking of that, but I'll probably make another PR in the future for that. Shouldn't be complicated, though.

LTVA1 commented:
Yep, when loading sample or wavetable

yohannd1 committed:
ffmpeg export: "fixing" 16-bit PCM output

yohannd1 committed:
ffmpeg export: m4a & opus support

yohannd1 committed:
ffmpeg export: improved export command system

Each ffmpeg export command is now specified by a template string, and
this opens up a window for future proper custom commands.

yohannd1 commented:
Progress update: haven't worked on this for a few weeks (carried away with a few other projects...), but so far the exporting is working.
I've hit kind of a roadblock with a bug though: the exported result is way faster than the song's intended speed. I'd be happy if someone could take a look cuz I think it needs a pair of fresh eyes.

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
ffmpeg export: fix issue with cuts in exporting! (hacky)

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
ffmpeg export: auto-detect ffmpeg path

yohannd1 committed:
ffmpeg export: LE optimization

yohannd1 committed:
ffmpeg export: improve BE-LE byte swap

yohannd1 committed:
ffmpeg export: proper path searching on windows (unfinished)

I forgot to commit this. I think I did this 2 weeks ago? 3?

yohannd1 committed:
ffmpeg export: the logging works on wine...

It's just the buffering that was odd.

yohannd1 committed:
ffmpeg export: add ffmpeg path setting

yohannd1 committed:
ffmpeg export: make vec2str into a proper function

yohannd1 committed:
ffmpeg export: support many_sys export mode (very incomplete)

yohannd1 committed:
ffmpeg export: remove exportOutputs arg from writeFloat/writeShort

It's much better being in the struct itself (and for SndfileWavWriter
the sndfile function itself already handles it).

yohannd1 committed:
ffmpeg export: move doExport lambda out of the switch statement

yohannd1 committed:
ffmpeg export: move many_sys export into lambda

yohannd1 committed:
ffmpeg export: simplify many_sys lambda

yohannd1 committed:
ffmpeg export: trying to make ffmpeg work on many_sys

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into 202504-ffmpeg-export-many_sys-attempt2

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into 202504-ffmpeg-export-many_sys-attempt2

yohannd1 committed:
Revert "ffmpeg export: remove exportOutputs arg from writeFloat/writeShort"

This reverts commit 26f8cb598c9228aa7d104b8028a7e07ae7a47504.

yohannd1 committed:
Revert "ffmpeg export: support many_sys export mode (very incomplete)"

This reverts commit d94167f468cb00240b29094df12e6175bab53eaf.

yohannd1 committed:
Merge branch '202504-ffmpeg-export-many_sys-attempt2' into ffmpeg-export

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
Merge branch 'master' of https://github.com/tildearrow/furnace into ffmpeg-export

yohannd1 committed:
ffmpeg export: try to make multi-chip work...

Not working yet, but I think I'm closer now.

yohannd1 committed:
ffmpeg export: small refactor

Apparently this fixed the multichip export - what?????????

yohannd1 committed:
ffmpeg export: fix read pointer going back when pipe blocks

yohannd1 committed:
ffmpeg export: use heap buffers instead of stack-allocated ones

Because C++ doesn't have dynamic stack allocation for some reason...

yohannd1 committed:
ffmpeg export: thanks MacOS CI compiler you are smarter than me

yohannd1 committed:
ffmpeg export: improve code doc

yohannd1 committed:
ffmpeg export: begin multi-chip export

yohannd1 committed:
ffmpeg export: finish multi-chip export

The effort here was trivial compared to what I had to endure in the
other two versions, but I think it's because it doesn't change much, so
I didn't have to change much either.
